crates/codegen/parser/src/char_set.rs

```diff
@@ -0,0 +1,251 @@
+                    if range.start == 0 {
+                        let end = unsafe { char::from_u32_unchecked(range.end - 1) };
+                        quote! { c <= #end }
+                    } else if range.end == u32::MAX {
+                        let start = unsafe { char::from_u32_unchecked(range.start) };
+                        quote! { #start <= c }
+                    } else if range.start + 1 == range.end {
+                        let start = unsafe { char::from_u32_unchecked(range.start) };
+                        quote! { c == #start }
+                    } else {
+                        let start = unsafe { char::from_u32_unchecked(range.start) };
+                        let end = unsafe { char::from_u32_unchecked(range.end - 1) };
+                        quote! { (#start <= c && c <= #end) }
+                    }
+                });
+                quote! { #(#clauses)||* }
+            }
+        }
+    }
+
+    pub fn from_expression(
+        tree: &CombinatorTree<'_>,
+        expression: &ExpressionRef,
+        inline_references: bool,
+    ) -> Option<CharSet> {
+        match &expression.parser {
+            ExpressionParser::Not(child) => {
+                Self::from_expression(tree, child, inline_references).map(Self::not)
+            }
+
+            ExpressionParser::Choice(children) => {
+                children.iter().fold(Some(Self::empty()), |acc, child| {
+                    acc.and_then(|acc| {
+                        Self::from_expression(tree, child, inline_references)
+                            .map(|child| acc.union(child))
+                    })
+                })
+            }
+
+            ExpressionParser::Terminal(string) => {
+                if string.chars().count() == 1 {
+                    Some(Self::single(string.chars().next().unwrap()))
+                } else {
+                    None
+                }
+            }
+
+            ExpressionParser::Range { from, to } => Some(Self::range(*from, *to)),
+
+            ExpressionParser::Difference {
+                minuend,
+                subtrahend,
+            } => {
+                if let (Some(minuend), Some(subtrahend)) = (
+                    Self::from_expression(tree, minuend, inline_references),
+                    Self::from_expression(tree, subtrahend, inline_references),
+                ) {
+                    Some(minuend.intersection(subtrahend.not()))
+                } else {
+                    None
+                }
+            }
+
+            ExpressionParser::Reference(name) => {
+                if inline_references {
+                    Self::from_expression(
+                        tree,
+                        &tree.context.get_tree_by_name(name).expression(),
+                        inline_references,
+                    )
+                } else {
+                    None
+                }
+            }
+
+            _ => None,
```

Comment:
Should it fail/error on unexpected variants instead of returning `None`? Is there a use case I'm missing?